### PR TITLE
PS-6865: Merge fb-prod201903 (5.7)

### PR DESCRIPTION
**Pull request by inikep**
1. Update rocksdb submodule to [8a1ecd1982341cfe073924d36717e11446cbe492 Fix build failures due to missing JEMALLOC_CXX_THROW macro (#5053)](https://github.com/facebook/rocksdb/commit/8a1ecd1982341cfe073924d36717e11446cbe492) (v5.8-1365-g8a1ecd198)

2. Add new MyRocks variables `read_free_rpl` and `enable_insert_with_update_caching`.

3. Cherry-pick the following commits from `fb-prod201903`:
- [da9bd74f8d5752cf83a84a655a6d3eb38152180c Added a new variable to control read-free replication](https://github.com/facebook/mysql-5.6/commit/da9bd74f8d5752cf83a84a655a6d3eb38152180c)
- [18b026729d68099f50aae9b40413b871646bc229 re-record the result of mtr test rocksdb.show_engine](https://github.com/facebook/mysql-5.6/commit/18b026729d68099f50aae9b40413b871646bc229)
- [c470e31144ffec8ba228b4272ce378e2ed4192d1 Update key fields encode/decode into static func](https://github.com/facebook/mysql-5.6/commit/c470e31144ffec8ba228b4272ce378e2ed4192d1)
- [e1516e7601e5a721459ab38054a8d75091736f41 Add new class Rdb_converter to expose decode API](https://github.com/facebook/mysql-5.6/commit/e1516e7601e5a721459ab38054a8d75091736f41)
- [77032004ad23d21a4c386f8136ecfbb071ea42d6 Simplify TTL value fetch for ttl column](https://github.com/facebook/mysql-5.6/commit/77032004ad23d21a4c386f8136ecfbb071ea42d6)
- [ab742e469d73f42251706a078c6b4506b526b3f1 Add encode value slice into Rdb_converter](https://github.com/facebook/mysql-5.6/commit/ab742e469d73f42251706a078c6b4506b526b3f1)
- [a6d547e4e6ac3e642173c20b5dc990df829006c9 Check thd flags during query processing](https://github.com/facebook/mysql-5.6/commit/a6d547e4e6ac3e642173c20b5dc990df829006c9)
- [ef4e62a6b04ca15eac95627210fcb51336e4fcf5 Add iterator class for decoding value slice](https://github.com/facebook/mysql-5.6/commit/ef4e62a6b04ca15eac95627210fcb51336e4fcf5)
- [469134bea110a7f1c888eb6ebbe4bce5a85b69be Avoid 2nd GetForUpdate on INSERT on DUPLICATE KEY. Improves performance up to 12.98%.](https://github.com/facebook/mysql-5.6/commit/469134bea110a7f1c888eb6ebbe4bce5a85b69be)

4. Null cherry-picked upstream commits:
- [51dbcc8da12bd28eb1510551eb228c82490cd23c Modifying histogram label for last bucket](https://github.com/facebook/mysql-5.6/commit/51dbcc8da12bd28eb1510551eb228c82490cd23c)
- [309f324ea3238006e9f2c0cadfc62b93de461dae Add rocksdb_bulk_load_allow_sk option in mysqldump - this enables bulk loading secondary keys in mysqldump output](https://github.com/facebook/mysql-5.6/commit/309f324ea3238006e9f2c0cadfc62b93de461dae)
- [f7d9a8dd3772212458c428924f04e5dcf9ddd8e0 Bug#84958 - InnoDB's MVCC has O(N^2) behaviors (#975)](https://github.com/facebook/mysql-5.6/commit/f7d9a8dd3772212458c428924f04e5dcf9ddd8e0)
- [927849171bf264dfb865bcfc7b586c51bc6722c0 Fix the gtid used during idempotent recovery](https://github.com/facebook/mysql-5.6/commit/927849171bf264dfb865bcfc7b586c51bc6722c0)
- [51a625078b58ef2d17ac1ce78af40bb30a9729b5 Sync transaction when number of keys exceed a threshold](https://github.com/facebook/mysql-5.6/commit/51a625078b58ef2d17ac1ce78af40bb30a9729b5)
- [75017433daa20546f9a71f4590e2ee337efed0a7 Optimize replace into to do a blind write](https://github.com/facebook/mysql-5.6/commit/75017433daa20546f9a71f4590e2ee337efed0a7)
- [98b0d40af27051ded465a5ca81fac8e3110cc2c1 Fix error handling in group by plans](https://github.com/facebook/mysql-5.6/commit/98b0d40af27051ded465a5ca81fac8e3110cc2c1)
- [7f81d6f82f3ec68267527a2fd6e774a79deae2b0 Backport HA_ERR_QUERY_INTERRUPTED to 5.6](https://github.com/facebook/mysql-5.6/commit/7f81d6f82f3ec68267527a2fd6e774a79deae2b0)
- [705104c1fd4a536d90f0fec481b06165ef41eea6 Added -DZSTD_STATIC_LINKING_ONLY option](https://github.com/facebook/mysql-5.6/commit/705104c1fd4a536d90f0fec481b06165ef41eea6)
- [0728f653551aa7784bfc780bbe8616e5f7bc12b9 Fix a optimized replace into test.](https://github.com/facebook/mysql-5.6/commit/0728f653551aa7784bfc780bbe8616e5f7bc12b9)
- [d2329cc28b64dab090a2358662faf87a66d9520b Handle column type conversion during key parsing](https://github.com/facebook/mysql-5.6/commit/d2329cc28b64dab090a2358662faf87a66d9520b)

### CODE REVIEW

File: storage/rocksdb/rdb_converter.cc [hunk 1]
```diff
@@ -0,0 +1,837 @@
...
+#include "./rdb_converter.h"
+
+/* Standard C++ header files */
+#include <algorithm>
+#include <map>
+#include <string>
+#include <vector>
+
+/* MySQL header files */
+#include "./log.h"
```
Comment (george-lorch) [inline storage/rocksdb/rdb_converter.cc]: In prep for 8.0 and in keeping with 5.7, we don't use "./xxx.h" idiom anymore for MySQL includes, just use "xxx.h" or more specific in 8.0 will be "sql/xxx.h".  see https://github.com/percona/percona-server/blob/50930fb400d882d432e9b2189fb95b3ed04356d3/storage/rocksdb/rdb_index_merge.h#L20 as an example.

File: storage/rocksdb/rdb_converter.cc [hunk 2]
```diff
@@ -0,0 +1,837 @@
...
+#include "./my_stacktrace.h"
+#include "./sql_array.h"
+
+/* MyRocks header files */
+#include "./ha_rocksdb.h"
```
Comment (george-lorch) [inline storage/rocksdb/rdb_converter.cc:32]: These are OK with the "./xxx.h as it is MyRocks convention.

File: storage/rocksdb/rdb_converter.h [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+  // get current field type
+  enum_field_types get_field_type() const;
+  // get current field
+  Field *get_field() const;
+};
+
+/**
+  Class to convert Mysql formats to rocksdb storage format, and vice versa.
+*/
+class Rdb_converter {
```
Comment (george-lorch) [inline storage/rocksdb/rdb_converter.h:128]: Please review these class methods for const correctness.  They may all be OK, but just please place a close eye on it.

**Review by george-lorch — APPROVED** [review]

**Comment by inikep** [comment]
Tested at https://ps57.cd.percona.com/job/percona-server-5.7-param/242

**Pull request merged**